From a741ff2557f7efe4c8f0b3acb7b6fefe07b26231 Mon Sep 17 00:00:00 2001
From: Hao Bui <hao.bui.yg@renesas.com>
Date: Wed, 27 Jan 2021 14:03:12 +0700
Subject: [PATCH 430/433] ARM: dts: iwg22d: Restructure devicetree with iWave
 RZ/G1E R4.0 support

Restructure devicetree and add basic support for the iWave RZ/G1E R4.0
board.

Signed-off-by: Hao Bui <hao.bui.yg@renesas.com>
---
 arch/arm/boot/dts/iwg22d-q7-common.dtsi            | 441 +++++++++++++++++++++
 arch/arm/boot/dts/iwg22d-q7-dbcm-ca.dtsi           | 165 ++++++++
 .../dts/r8a7745-iwg22d-sodimm-dbhd-ca-rev4.dts     |  26 ++
 .../arm/boot/dts/r8a7745-iwg22d-sodimm-dbhd-ca.dts | 159 +-------
 arch/arm/boot/dts/r8a7745-iwg22d-sodimm-rev4.dts   |  49 +++
 arch/arm/boot/dts/r8a7745-iwg22d-sodimm.dts        | 409 +------------------
 arch/arm/boot/dts/r8a7745-iwg22m-rev4.dtsi         |  36 ++
 7 files changed, 724 insertions(+), 561 deletions(-)
 create mode 100644 arch/arm/boot/dts/iwg22d-q7-common.dtsi
 create mode 100644 arch/arm/boot/dts/iwg22d-q7-dbcm-ca.dtsi
 create mode 100644 arch/arm/boot/dts/r8a7745-iwg22d-sodimm-dbhd-ca-rev4.dts
 create mode 100644 arch/arm/boot/dts/r8a7745-iwg22d-sodimm-rev4.dts
 create mode 100644 arch/arm/boot/dts/r8a7745-iwg22m-rev4.dtsi

diff --git a/arch/arm/boot/dts/iwg22d-q7-common.dtsi b/arch/arm/boot/dts/iwg22d-q7-common.dtsi
new file mode 100644
index 0000000..03389f7
--- /dev/null
+++ b/arch/arm/boot/dts/iwg22d-q7-common.dtsi
@@ -0,0 +1,441 @@
+/*
+ * Device Tree Source for the iWave-RZG1E SODIMM carrier board
+ *
+ * Copyright (C) 2017 Renesas Electronics Corp.
+ *
+ * This file is licensed under the terms of the GNU General Public License
+ * version 2.  This program is licensed "as is" without any warranty of any
+ * kind, whether express or implied.
+ */
+
+/*
+ * SSI-SGTL5000
+ *
+ * This command is required when Playback/Capture
+ *
+ *      amixer set "DVC Out" 100%
+ *      amixer set "DVC In" 100%
+ *
+ * You can use Mute
+ *
+ *      amixer set "DVC Out Mute" on
+ *      amixer set "DVC In Mute" on
+ *
+ * You can use Volume Ramp
+ *
+ *      amixer set "DVC Out Ramp Up Rate"   "0.125 dB/64 steps"
+ *      amixer set "DVC Out Ramp Down Rate" "0.125 dB/512 steps"
+ *      amixer set "DVC Out Ramp" on
+ *      aplay xxx.wav &
+ *      amixer set "DVC Out"  80%  // Volume Down
+ *      amixer set "DVC Out" 100%  // Volume Up
+ */
+
+#include <dt-bindings/pwm/pwm.h>
+
+/ {
+	aliases {
+		serial1 = &scif2;
+		serial2 = &scif3;
+		serial3 = &scif4;
+		serial5 = &hscif1;
+		ethernet0 = &avb;
+	};
+
+	chosen {
+		bootargs = "ignore_loglevel rw root=/dev/nfs ip=dhcp";
+		stdout-path = "serial3:115200n8";
+	};
+
+	audio_clock: audio_clock {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-frequency = <26000000>;
+	};
+
+	rsnd_sgtl5000: sound {
+		compatible = "simple-audio-card";
+		simple-audio-card,format = "i2s";
+		simple-audio-card,bitclock-master = <&sndcodec>;
+		simple-audio-card,frame-master = <&sndcodec>;
+
+		sndcpu: simple-audio-card,cpu {
+			sound-dai = <&rcar_sound>;
+		};
+
+		sndcodec: simple-audio-card,codec {
+			sound-dai = <&sgtl5000>;
+		};
+	};
+
+	vccq_sdhi0: regulator-vccq-sdhi0 {
+		compatible = "regulator-gpio";
+
+		regulator-name = "SDHI0 VccQ";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <3300000>;
+
+		gpios = <&gpio0 20 GPIO_ACTIVE_LOW>;
+		gpios-states = <1>;
+		states = <3300000 1
+			  1800000 0>;
+	};
+
+	vccq_panel: regulator-vccq-panel {
+		compatible = "regulator-fixed";
+		regulator-name = "Panel VccQ";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		gpio = <&gpio1 13 GPIO_ACTIVE_LOW>;
+	};
+
+	backlight_lcd: backlight {
+		compatible = "pwm-backlight";
+		pwms = <&tpu 3 5000000 PWM_POLARITY_INVERTED>;
+		brightness-levels = <0 4 8 16 32 64 128 255>;
+		default-brightness-level = <7>;
+	};
+
+	lcd_panel: lcd {
+		compatible = "edt,etm043080dh6gp";
+		power-supply = <&vccq_panel>;
+		backlight = <&backlight_lcd>;
+
+		port {
+			lcd_in: endpoint {
+				remote-endpoint = <&du_out_rgb0>;
+			};
+		};
+	};
+};
+
+&du {
+	pinctrl-0 = <&du0_pins>;
+	pinctrl-names = "default";
+
+	status = "okay";
+
+	ports {
+		port@0 {
+			endpoint {
+				remote-endpoint = <&lcd_in>;
+			};
+		};
+	};
+};
+
+&can0 {
+	pinctrl-0 = <&can0_pins>;
+	pinctrl-names = "default";
+
+	status = "okay";
+};
+
+&gpio4 {
+	can0-trx-en-gpio {
+		gpio-hog;
+		gpios = <2 GPIO_ACTIVE_HIGH>;
+		output-low;
+		line-name = "can0-trx-en-gpio";
+	};
+
+	touch-interrupt {
+		gpio-hog;
+		gpios = <4 GPIO_ACTIVE_LOW>;
+		input;
+	};
+};
+
+&hscif1 {
+	pinctrl-0 = <&hscif1_pins>;
+	pinctrl-names = "default";
+
+	uart-has-rtscts;
+	status = "okay";
+};
+
+&i2c5 {
+	pinctrl-0 = <&i2c5_pins>;
+	pinctrl-names = "default";
+
+	status = "okay";
+	clock-frequency = <400000>;
+
+	stmpe811@44 {
+		compatible = "st,stmpe811";
+		reg = <0x44>;
+		interrupt-parent = <&gpio4>;
+		interrupts = <4 IRQ_TYPE_LEVEL_LOW>;
+
+		/* 3.25 MHz ADC clock speed */
+		st,adc-freq = <1>;
+		/* ADC converstion time: 80 clocks */
+		st,sample-time = <4>;
+		/* 12-bit ADC */
+		st,mod-12b = <1>;
+		/* internal ADC reference */
+		st,ref-sel = <0>;
+
+		stmpe_touchscreen {
+			compatible = "st,stmpe-ts";
+			/* 8 sample average control */
+			st,ave-ctrl = <3>;
+			/* 7 length fractional part in z */
+			st,fraction-z = <7>;
+			/*
+			 * 50 mA typical 80 mA max touchscreen drivers
+			 * current limit value
+			 */
+			st,i-drive = <1>;
+			/* 1 ms panel driver settling time */
+			st,settling = <3>;
+			/* 5 ms touch detect interrupt delay */
+			st,touch-det-delay = <5>;
+		};
+	};
+
+	sgtl5000: codec@a {
+		compatible = "fsl,sgtl5000";
+		#sound-dai-cells = <0>;
+		reg = <0x0a>;
+		clocks = <&audio_clock>;
+		VDDA-supply = <&reg_3p3v>;
+		VDDIO-supply = <&reg_3p3v>;
+	};
+
+	ov5640@3c {
+		compatible = "ovti,ov5640";
+		reg = <0x3c>;
+		remote = <&vin1>;
+
+		port {
+			ov5640: endpoint {
+				bus-width = <8>;
+				pclk-sample = <1>;
+				remote-endpoint = <&vin1ep>;
+			};
+		};
+	};
+
+	ov7725@21 {
+		compatible = "ov772x";
+		reg = <0x21>;
+		remote = <&vin1>;
+
+		/* mark OV7725 as disabled by default
+		 * to enable, run command below under U-boot command line
+		 * 	setenv vin1_camera 'ov7725'
+		 * Note: no need to update status here
+		 */
+		status = "disabled";
+
+		port {
+			ov7725: endpoint {
+				bus-width = <8>;
+				remote-endpoint = <&vin1ep>;
+			};
+		};
+	};
+};
+
+&pci0 {
+	status = "okay";
+};
+
+&pfc {
+	backlight_pins: backlight {
+		groups = "tpu_to3_c";
+		function = "tpu";
+	};
+
+	can0_pins: can0 {
+		groups = "can0_data";
+		function = "can0";
+	};
+
+	du0_pins: du0 {
+		groups = "du0_rgb666", "du0_sync", "du0_disp", "du0_clk0_out";
+		function = "du0";
+	};
+
+	hscif1_pins: hscif1 {
+		groups = "hscif1_data", "hscif1_ctrl";
+		function = "hscif1";
+	};
+
+	i2c5_pins: i2c5 {
+		groups = "i2c5_b";
+		function = "i2c5";
+	};
+
+	scif2_pins: serial1 {
+		groups = "scif2_data";
+		function = "scif2";
+	};
+
+	scif3_pins: serial2 {
+		groups = "scif3_data";
+		function = "scif3";
+	};
+
+	scif4_pins: scif4 {
+		groups = "scif4_data_b";
+		function = "scif4";
+	};
+
+	avb_pins: avb {
+		groups = "avb_mdio", "avb_gmii";
+		function = "avb";
+	};
+
+	sdhi0_pins: sd0 {
+		groups = "sdhi0_data4", "sdhi0_ctrl";
+		function = "sdhi0";
+		power-source = <3300>;
+	};
+
+	sdhi0_pins_uhs: sd0_uhs {
+		groups = "sdhi0_data4", "sdhi0_ctrl";
+		function = "sdhi0";
+		power-source = <1800>;
+	};
+
+	sound_pins: sound {
+		groups = "ssi34_ctrl", "ssi3_data", "ssi4_data";
+		function = "ssi";
+	};
+
+	usb0_pins: usb0 {
+		groups = "usb0";
+		function = "usb0";
+	};
+
+	usb1_pins: usb1 {
+		groups = "usb1";
+		function = "usb1";
+	};
+
+	vin1_pins: vin1 {
+		groups = "vin1_data8", "vin1_sync", "vin1_clk";
+		function = "vin1";
+	};
+};
+
+&rcar_sound {
+	pinctrl-0 = <&sound_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+
+	/* Single DAI */
+
+	#sound-dai-cells = <0>;
+
+	rcar_sound,dai {
+		dai0 {
+			playback = <&ssi3 &src3 &dvc0>;
+			capture = <&ssi4 &src4 &dvc1>;
+		};
+	};
+};
+
+&scif2 {
+	pinctrl-0 = <&scif2_pins>;
+	pinctrl-names = "default";
+
+	status = "okay";
+};
+
+&scif3 {
+	pinctrl-0 = <&scif3_pins>;
+	pinctrl-names = "default";
+
+	status = "okay";
+};
+
+&scif4 {
+	pinctrl-0 = <&scif4_pins>;
+	pinctrl-names = "default";
+
+	status = "okay";
+};
+
+&avb {
+	pinctrl-0 = <&avb_pins>;
+	pinctrl-names = "default";
+
+	phy-handle = <&phy3>;
+	phy-mode = "gmii";
+	renesas,no-ether-link;
+	status = "okay";
+
+	phy3: ethernet-phy@3 {
+	/*
+	 * On some older versions of the platform (before R4.0) the phy address
+	 * may be 1 or 3. The address is fixed to 3 for R4.0 onwards.
+	 * Remove <reg> property to enable autoscan function
+	 */
+		/* reg = <1>; */
+		micrel,led-mode = <1>;
+	};
+};
+
+&tmu1 {
+	status = "okay";
+};
+
+&sdhi0 {
+	pinctrl-0 = <&sdhi0_pins>;
+	pinctrl-1 = <&sdhi0_pins_uhs>;
+	pinctrl-names = "default", "state_uhs";
+
+	vmmc-supply = <&reg_3p3v>;
+	vqmmc-supply = <&vccq_sdhi0>;
+	cd-gpios = <&gpio6 6 GPIO_ACTIVE_LOW>;
+	status = "okay";
+};
+
+&hsusb {
+	status = "okay";
+	pinctrl-0 = <&usb0_pins>;
+	pinctrl-names = "default";
+};
+
+&pci1 {
+	status = "okay";
+	pinctrl-0 = <&usb1_pins>;
+	pinctrl-names = "default";
+};
+
+&ssi4 {
+	shared-pin;
+};
+
+&tpu {
+	pinctrl-0 = <&backlight_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+};
+
+&usbphy {
+	status = "okay";
+	renesas,id-gpio = <&gpio4 6 GPIO_ACTIVE_LOW>;
+	renesas,vbus-gpio = <&gpio5 25 GPIO_ACTIVE_LOW>;
+	renesas,vbus-pwr-gpio = <&gpio5 24 GPIO_ACTIVE_LOW>;
+	renesas,wakeup;
+};
+
+&vin1 {
+	status = "okay";
+	pinctrl-0 = <&vin1_pins>;
+	pinctrl-names = "default";
+
+	port {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		vin1ep: endpoint {
+			remote-endpoint = <&ov5640>;
+			bus-width = <8>;
+		};
+	};
+};
diff --git a/arch/arm/boot/dts/iwg22d-q7-dbcm-ca.dtsi b/arch/arm/boot/dts/iwg22d-q7-dbcm-ca.dtsi
new file mode 100644
index 0000000..b3ccd4b
--- /dev/null
+++ b/arch/arm/boot/dts/iwg22d-q7-dbcm-ca.dtsi
@@ -0,0 +1,165 @@
+/*
+ * Device Tree Source for the iWave-RZ/G1E Daughter Board Camera Module
+ *
+ * Copyright (C) 2021 Renesas Electronics Corp.
+ *
+ * This file is licensed under the terms of the GNU General Public License
+ * version 2.  This program is licensed "as is" without any warranty of any
+ * kind, whether express or implied.
+ */
+
+/ {
+	aliases {
+		serial0 = &scif1;
+		serial4 = &scif5;
+		serial6 = &hscif2;
+	};
+
+	cec_clock: cec-clock {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-frequency = <12000000>;
+	};
+
+	hdmi-out {
+		compatible = "hdmi-connector";
+		type = "a";
+
+		port {
+			hdmi_con: endpoint {
+				remote-endpoint = <&adv7511_out>;
+			};
+		};
+	};
+};
+
+&du {
+	pinctrl-0 = <&du0_pins>;
+	pinctrl-names = "default";
+
+	status = "okay";
+
+	ports {
+		port@0 {
+			endpoint {
+				remote-endpoint = <&adv7511_in>;
+			};
+		};
+	};
+};
+
+&can1 {
+	pinctrl-0 = <&can1_pins>;
+	pinctrl-names = "default";
+
+	status = "okay";
+};
+
+&gpio4 {
+	can1-trx-en-gpio {
+		gpio-hog;
+		gpios = <10 GPIO_ACTIVE_HIGH>;
+		output-low;
+		line-name = "can1-trx-en-gpio";
+	};
+};
+
+&hscif2 {
+	pinctrl-0 = <&hscif2_pins>;
+	pinctrl-names = "default";
+
+	status = "okay";
+};
+
+&i2c1 {
+	pinctrl-0 = <&i2c1_pins>;
+	pinctrl-names = "default";
+
+	status = "okay";
+	clock-frequency = <400000>;
+
+	hdmi@39 {
+		compatible = "adi,adv7511w";
+		reg = <0x39>;
+		clocks = <&cec_clock>;
+		clock-names = "cec";
+		pd-gpios = <&gpio2 24 GPIO_ACTIVE_HIGH>;
+
+		adi,input-depth = <8>;
+		adi,input-colorspace = "rgb";
+		adi,input-clock = "1x";
+		adi,input-style = <1>;
+		adi,input-justification = "evenly";
+
+		ports {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			port@0 {
+				reg = <0>;
+				adv7511_in: endpoint {
+					remote-endpoint = <&du_out_rgb0>;
+				};
+			};
+
+			port@1 {
+				reg = <1>;
+				adv7511_out: endpoint {
+					remote-endpoint = <&hdmi_con>;
+				};
+			};
+		};
+	};
+};
+
+&lcd_panel {
+	status = "disabled";
+
+	/delete-node/ port;
+};
+
+&pfc {
+	can1_pins: can1 {
+		groups = "can1_data_b";
+		function = "can1";
+	};
+
+	du0_pins: du0 {
+		groups = "du0_rgb888", "du0_sync", "du0_disp", "du0_clk0_out";
+		function = "du0";
+	};
+
+	hscif2_pins: hscif2 {
+		groups = "hscif2_data";
+		function = "hscif2";
+	};
+
+	i2c1_pins: i2c1 {
+		groups = "i2c1_d";
+		function = "i2c1";
+	};
+
+	scif1_pins: scif1 {
+		groups = "scif1_data";
+		function = "scif1";
+	};
+
+	scif5_pins: scif5 {
+		groups = "scif5_data_d";
+		function = "scif5";
+	};
+};
+
+&scif1 {
+	pinctrl-0 = <&scif1_pins>;
+	pinctrl-names = "default";
+
+	status = "okay";
+};
+
+&scif5 {
+	pinctrl-0 = <&scif5_pins>;
+	pinctrl-names = "default";
+
+	status = "okay";
+};
diff --git a/arch/arm/boot/dts/r8a7745-iwg22d-sodimm-dbhd-ca-rev4.dts b/arch/arm/boot/dts/r8a7745-iwg22d-sodimm-dbhd-ca-rev4.dts
new file mode 100644
index 0000000..033fc56
--- /dev/null
+++ b/arch/arm/boot/dts/r8a7745-iwg22d-sodimm-dbhd-ca-rev4.dts
@@ -0,0 +1,26 @@
+/*
+ * Device Tree Source for the iWave-RZG1E SODIMM carrier board + HDMI daughter
+ * board
+ *
+ * Copyright (C) 2017 Renesas Electronics Corp.
+ *
+ * This file is licensed under the terms of the GNU General Public License
+ * version 2.  This program is licensed "as is" without any warranty of any
+ * kind, whether express or implied.
+ */
+
+/dts-v1/;
+#include "r8a7745-iwg22m-rev4.dtsi"
+#include "iwg22d-q7-common.dtsi"
+#include "iwg22d-q7-dbcm-ca.dtsi"
+
+/ {
+	model = "iWave RainboW-G22D-SODIMM RZ/G1E based board R4.0 with HDMI add-on";
+	compatible = "iwave,g22d", "iwave,g22m", "renesas,r8a7745";
+};
+
+&avb {
+	phy3: ethernet-phy@3 {
+		reg = <3>;
+	};
+};
diff --git a/arch/arm/boot/dts/r8a7745-iwg22d-sodimm-dbhd-ca.dts b/arch/arm/boot/dts/r8a7745-iwg22d-sodimm-dbhd-ca.dts
index 46416ce..4034ade 100644
--- a/arch/arm/boot/dts/r8a7745-iwg22d-sodimm-dbhd-ca.dts
+++ b/arch/arm/boot/dts/r8a7745-iwg22d-sodimm-dbhd-ca.dts
@@ -9,163 +9,12 @@
  * kind, whether express or implied.
  */
 
-#include "r8a7745-iwg22d-sodimm.dts"
+/dts-v1/;
+#include "r8a7745-iwg22m.dtsi"
+#include "iwg22d-q7-common.dtsi"
+#include "iwg22d-q7-dbcm-ca.dtsi"
 
 / {
 	model = "iWave RainboW-G22D-SODIMM RZ/G1E based board with HDMI add-on";
 	compatible = "iwave,g22d", "iwave,g22m", "renesas,r8a7745";
-
-	aliases {
-		serial0 = &scif1;
-		serial4 = &scif5;
-		serial6 = &hscif2;
-	};
-
-	cec_clock: cec-clock {
-		compatible = "fixed-clock";
-		#clock-cells = <0>;
-		clock-frequency = <12000000>;
-	};
-
-	hdmi-out {
-		compatible = "hdmi-connector";
-		type = "a";
-
-		port {
-			hdmi_con: endpoint {
-				remote-endpoint = <&adv7511_out>;
-			};
-		};
-	};
-};
-
-&du {
-	pinctrl-0 = <&du0_pins>;
-	pinctrl-names = "default";
-
-	status = "okay";
-
-	ports {
-		port@0 {
-			endpoint {
-				remote-endpoint = <&adv7511_in>;
-			};
-		};
-	};
-};
-
-&can1 {
-	pinctrl-0 = <&can1_pins>;
-	pinctrl-names = "default";
-
-	status = "okay";
-};
-
-&gpio4 {
-	can1-trx-en-gpio {
-		gpio-hog;
-		gpios = <10 GPIO_ACTIVE_HIGH>;
-		output-low;
-		line-name = "can1-trx-en-gpio";
-	};
-};
-
-&hscif2 {
-	pinctrl-0 = <&hscif2_pins>;
-	pinctrl-names = "default";
-
-	status = "okay";
-};
-
-&i2c1 {
-	pinctrl-0 = <&i2c1_pins>;
-	pinctrl-names = "default";
-
-	status = "okay";
-	clock-frequency = <400000>;
-
-	hdmi@39 {
-		compatible = "adi,adv7511w";
-		reg = <0x39>;
-		clocks = <&cec_clock>;
-		clock-names = "cec";
-		pd-gpios = <&gpio2 24 GPIO_ACTIVE_HIGH>;
-
-		adi,input-depth = <8>;
-		adi,input-colorspace = "rgb";
-		adi,input-clock = "1x";
-		adi,input-style = <1>;
-		adi,input-justification = "evenly";
-
-		ports {
-			#address-cells = <1>;
-			#size-cells = <0>;
-
-			port@0 {
-				reg = <0>;
-				adv7511_in: endpoint {
-					remote-endpoint = <&du_out_rgb0>;
-				};
-			};
-
-			port@1 {
-				reg = <1>;
-				adv7511_out: endpoint {
-					remote-endpoint = <&hdmi_con>;
-				};
-			};
-		};
-	};
-};
-
-&lcd_panel {
-	status = "disabled";
-
-	/delete-node/ port;
-};
-
-&pfc {
-	can1_pins: can1 {
-		groups = "can1_data_b";
-		function = "can1";
-	};
-
-	du0_pins: du0 {
-		groups = "du0_rgb888", "du0_sync", "du0_disp", "du0_clk0_out";
-		function = "du0";
-	};
-
-	hscif2_pins: hscif2 {
-		groups = "hscif2_data";
-		function = "hscif2";
-	};
-
-	i2c1_pins: i2c1 {
-		groups = "i2c1_d";
-		function = "i2c1";
-	};
-
-	scif1_pins: scif1 {
-		groups = "scif1_data";
-		function = "scif1";
-	};
-
-	scif5_pins: scif5 {
-		groups = "scif5_data_d";
-		function = "scif5";
-	};
-};
-
-&scif1 {
-	pinctrl-0 = <&scif1_pins>;
-	pinctrl-names = "default";
-
-	status = "okay";
-};
-
-&scif5 {
-	pinctrl-0 = <&scif5_pins>;
-	pinctrl-names = "default";
-
-	status = "okay";
 };
diff --git a/arch/arm/boot/dts/r8a7745-iwg22d-sodimm-rev4.dts b/arch/arm/boot/dts/r8a7745-iwg22d-sodimm-rev4.dts
new file mode 100644
index 0000000..635173c
--- /dev/null
+++ b/arch/arm/boot/dts/r8a7745-iwg22d-sodimm-rev4.dts
@@ -0,0 +1,49 @@
+/*
+ * Device Tree Source for the iWave-RZG1E SODIMM Development Platform
+ *  with R4.0 SOM
+ *
+ * Copyright (C) 2021 Renesas Electronics Corp.
+ *
+ * This file is licensed under the terms of the GNU General Public License
+ * version 2.  This program is licensed "as is" without any warranty of any
+ * kind, whether express or implied.
+ */
+
+/*
+ * SSI-SGTL5000
+ *
+ * This command is required when Playback/Capture
+ *
+ *      amixer set "DVC Out" 100%
+ *      amixer set "DVC In" 100%
+ *
+ * You can use Mute
+ *
+ *      amixer set "DVC Out Mute" on
+ *      amixer set "DVC In Mute" on
+ *
+ * You can use Volume Ramp
+ *
+ *      amixer set "DVC Out Ramp Up Rate"   "0.125 dB/64 steps"
+ *      amixer set "DVC Out Ramp Down Rate" "0.125 dB/512 steps"
+ *      amixer set "DVC Out Ramp" on
+ *      aplay xxx.wav &
+ *      amixer set "DVC Out"  80%  // Volume Down
+ *      amixer set "DVC Out" 100%  // Volume Up
+ */
+
+/dts-v1/;
+#include "r8a7745-iwg22m-rev4.dtsi"
+#include "iwg22d-q7-common.dtsi"
+
+/ {
+	model = "iWave Systems RainboW-G22D-SODIMM R4.0 board based on RZ/G1E";
+	compatible = "iwave,g22d", "iwave,g22m", "renesas,r8a7745";
+
+};
+
+&avb {
+	phy3: ethernet-phy@3 {
+		reg = <3>;
+	};
+};
diff --git a/arch/arm/boot/dts/r8a7745-iwg22d-sodimm.dts b/arch/arm/boot/dts/r8a7745-iwg22d-sodimm.dts
index e6eb7a1..9be0681 100644
--- a/arch/arm/boot/dts/r8a7745-iwg22d-sodimm.dts
+++ b/arch/arm/boot/dts/r8a7745-iwg22d-sodimm.dts
@@ -1,5 +1,6 @@
 /*
- * Device Tree Source for the iWave-RZG1E SODIMM carrier board
+ * Device Tree Source for the iWave-RZG1E SODIMM Development Platform
+ *  with R2.0 SOM
  *
  * Copyright (C) 2017 Renesas Electronics Corp.
  *
@@ -33,413 +34,9 @@
 
 /dts-v1/;
 #include "r8a7745-iwg22m.dtsi"
-#include <dt-bindings/pwm/pwm.h>
+#include "iwg22d-q7-common.dtsi"
 
 / {
 	model = "iWave Systems RainboW-G22D-SODIMM board based on RZ/G1E";
 	compatible = "iwave,g22d", "iwave,g22m", "renesas,r8a7745";
-
-	aliases {
-		serial1 = &scif2;
-		serial2 = &scif3;
-		serial3 = &scif4;
-		serial5 = &hscif1;
-		ethernet0 = &avb;
-	};
-
-	chosen {
-		bootargs = "ignore_loglevel rw root=/dev/nfs ip=dhcp";
-		stdout-path = "serial3:115200n8";
-	};
-
-	audio_clock: audio_clock {
-		compatible = "fixed-clock";
-		#clock-cells = <0>;
-		clock-frequency = <26000000>;
-	};
-
-	rsnd_sgtl5000: sound {
-		compatible = "simple-audio-card";
-		simple-audio-card,format = "i2s";
-		simple-audio-card,bitclock-master = <&sndcodec>;
-		simple-audio-card,frame-master = <&sndcodec>;
-
-		sndcpu: simple-audio-card,cpu {
-			sound-dai = <&rcar_sound>;
-		};
-
-		sndcodec: simple-audio-card,codec {
-			sound-dai = <&sgtl5000>;
-		};
-	};
-
-	vccq_sdhi0: regulator-vccq-sdhi0 {
-		compatible = "regulator-gpio";
-
-		regulator-name = "SDHI0 VccQ";
-		regulator-min-microvolt = <1800000>;
-		regulator-max-microvolt = <3300000>;
-
-		gpios = <&gpio0 20 GPIO_ACTIVE_LOW>;
-		gpios-states = <1>;
-		states = <3300000 1
-			  1800000 0>;
-	};
-
-	vccq_panel: regulator-vccq-panel {
-		compatible = "regulator-fixed";
-		regulator-name = "Panel VccQ";
-		regulator-min-microvolt = <3300000>;
-		regulator-max-microvolt = <3300000>;
-		gpio = <&gpio1 13 GPIO_ACTIVE_LOW>;
-	};
-
-	backlight_lcd: backlight {
-		compatible = "pwm-backlight";
-		pwms = <&tpu 3 5000000 PWM_POLARITY_INVERTED>;
-		brightness-levels = <0 4 8 16 32 64 128 255>;
-		default-brightness-level = <7>;
-	};
-
-	lcd_panel: lcd {
-		compatible = "edt,etm043080dh6gp";
-		power-supply = <&vccq_panel>;
-		backlight = <&backlight_lcd>;
-
-		port {
-			lcd_in: endpoint {
-				remote-endpoint = <&du_out_rgb0>;
-			};
-		};
-	};
-};
-
-&du {
-	pinctrl-0 = <&du0_pins>;
-	pinctrl-names = "default";
-
-	status = "okay";
-
-	ports {
-		port@0 {
-			endpoint {
-				remote-endpoint = <&lcd_in>;
-			};
-		};
-	};
-};
-
-&can0 {
-	pinctrl-0 = <&can0_pins>;
-	pinctrl-names = "default";
-
-	status = "okay";
-};
-
-&gpio4 {
-	can0-trx-en-gpio {
-		gpio-hog;
-		gpios = <2 GPIO_ACTIVE_HIGH>;
-		output-low;
-		line-name = "can0-trx-en-gpio";
-	};
-
-	touch-interrupt {
-		gpio-hog;
-		gpios = <4 GPIO_ACTIVE_LOW>;
-		input;
-	};
-};
-
-&hscif1 {
-	pinctrl-0 = <&hscif1_pins>;
-	pinctrl-names = "default";
-
-	uart-has-rtscts;
-	status = "okay";
-};
-
-&i2c5 {
-	pinctrl-0 = <&i2c5_pins>;
-	pinctrl-names = "default";
-
-	status = "okay";
-	clock-frequency = <400000>;
-
-	stmpe811@44 {
-		compatible = "st,stmpe811";
-		reg = <0x44>;
-		interrupt-parent = <&gpio4>;
-		interrupts = <4 IRQ_TYPE_LEVEL_LOW>;
-
-		/* 3.25 MHz ADC clock speed */
-		st,adc-freq = <1>;
-		/* ADC converstion time: 80 clocks */
-		st,sample-time = <4>;
-		/* 12-bit ADC */
-		st,mod-12b = <1>;
-		/* internal ADC reference */
-		st,ref-sel = <0>;
-
-		stmpe_touchscreen {
-			compatible = "st,stmpe-ts";
-			/* 8 sample average control */
-			st,ave-ctrl = <3>;
-			/* 7 length fractional part in z */
-			st,fraction-z = <7>;
-			/*
-			 * 50 mA typical 80 mA max touchscreen drivers
-			 * current limit value
-			 */
-			st,i-drive = <1>;
-			/* 1 ms panel driver settling time */
-			st,settling = <3>;
-			/* 5 ms touch detect interrupt delay */
-			st,touch-det-delay = <5>;
-		};
-	};
-
-	sgtl5000: codec@a {
-		compatible = "fsl,sgtl5000";
-		#sound-dai-cells = <0>;
-		reg = <0x0a>;
-		clocks = <&audio_clock>;
-		VDDA-supply = <&reg_3p3v>;
-		VDDIO-supply = <&reg_3p3v>;
-	};
-
-	ov5640@3c {
-		compatible = "ovti,ov5640";
-		reg = <0x3c>;
-		remote = <&vin1>;
-
-		port {
-			ov5640: endpoint {
-				bus-width = <8>;
-				pclk-sample = <1>;
-				remote-endpoint = <&vin1ep>;
-			};
-		};
-	};
-
-	ov7725@21 {
-		compatible = "ov772x";
-		reg = <0x21>;
-		remote = <&vin1>;
-
-		/* mark OV7725 as disabled by default
-		 * to enable, run command below under U-boot command line
-		 * 	setenv vin1_camera 'ov7725'
-		 * Note: no need to update status here
-		 */
-		status = "disabled";
-
-		port {
-			ov7725: endpoint {
-				bus-width = <8>;
-				remote-endpoint = <&vin1ep>;
-			};
-		};
-	};
-};
-
-&pci0 {
-	status = "okay";
-};
-
-&pfc {
-	backlight_pins: backlight {
-		groups = "tpu_to3_c";
-		function = "tpu";
-	};
-
-	can0_pins: can0 {
-		groups = "can0_data";
-		function = "can0";
-	};
-
-	du0_pins: du0 {
-		groups = "du0_rgb666", "du0_sync", "du0_disp", "du0_clk0_out";
-		function = "du0";
-	};
-
-	hscif1_pins: hscif1 {
-		groups = "hscif1_data", "hscif1_ctrl";
-		function = "hscif1";
-	};
-
-	i2c5_pins: i2c5 {
-		groups = "i2c5_b";
-		function = "i2c5";
-	};
-
-	scif2_pins: serial1 {
-		groups = "scif2_data";
-		function = "scif2";
-	};
-
-	scif3_pins: serial2 {
-		groups = "scif3_data";
-		function = "scif3";
-	};
-
-	scif4_pins: scif4 {
-		groups = "scif4_data_b";
-		function = "scif4";
-	};
-
-	avb_pins: avb {
-		groups = "avb_mdio", "avb_gmii";
-		function = "avb";
-	};
-
-	sdhi0_pins: sd0 {
-		groups = "sdhi0_data4", "sdhi0_ctrl";
-		function = "sdhi0";
-		power-source = <3300>;
-	};
-
-	sdhi0_pins_uhs: sd0_uhs {
-		groups = "sdhi0_data4", "sdhi0_ctrl";
-		function = "sdhi0";
-		power-source = <1800>;
-	};
-
-	sound_pins: sound {
-		groups = "ssi34_ctrl", "ssi3_data", "ssi4_data";
-		function = "ssi";
-	};
-
-	usb0_pins: usb0 {
-		groups = "usb0";
-		function = "usb0";
-	};
-
-	usb1_pins: usb1 {
-		groups = "usb1";
-		function = "usb1";
-	};
-
-	vin1_pins: vin1 {
-		groups = "vin1_data8", "vin1_sync", "vin1_clk";
-		function = "vin1";
-	};
-};
-
-&rcar_sound {
-	pinctrl-0 = <&sound_pins>;
-	pinctrl-names = "default";
-	status = "okay";
-
-	/* Single DAI */
-
-	#sound-dai-cells = <0>;
-
-	rcar_sound,dai {
-		dai0 {
-			playback = <&ssi3 &src3 &dvc0>;
-			capture = <&ssi4 &src4 &dvc1>;
-		};
-	};
-};
-
-&scif2 {
-	pinctrl-0 = <&scif2_pins>;
-	pinctrl-names = "default";
-
-	status = "okay";
-};
-
-&scif3 {
-	pinctrl-0 = <&scif3_pins>;
-	pinctrl-names = "default";
-
-	status = "okay";
-};
-
-&scif4 {
-	pinctrl-0 = <&scif4_pins>;
-	pinctrl-names = "default";
-
-	status = "okay";
-};
-
-&avb {
-	pinctrl-0 = <&avb_pins>;
-	pinctrl-names = "default";
-
-	phy-handle = <&phy3>;
-	phy-mode = "gmii";
-	renesas,no-ether-link;
-	status = "okay";
-
-	phy3: ethernet-phy@3 {
-	/*
-	 * On some older versions of the platform (before R4.0) the phy address
-	 * may be 1 or 3. The address is fixed to 3 for R4.0 onwards.
-	 */
-		reg = <3>;
-		micrel,led-mode = <1>;
-	};
-};
-
-&tmu1 {
-	status = "okay";
-};
-
-&sdhi0 {
-	pinctrl-0 = <&sdhi0_pins>;
-	pinctrl-1 = <&sdhi0_pins_uhs>;
-	pinctrl-names = "default", "state_uhs";
-
-	vmmc-supply = <&reg_3p3v>;
-	vqmmc-supply = <&vccq_sdhi0>;
-	cd-gpios = <&gpio6 6 GPIO_ACTIVE_LOW>;
-	status = "okay";
-};
-
-&hsusb {
-	status = "okay";
-	pinctrl-0 = <&usb0_pins>;
-	pinctrl-names = "default";
-};
-
-&pci1 {
-	status = "okay";
-	pinctrl-0 = <&usb1_pins>;
-	pinctrl-names = "default";
-};
-
-&ssi4 {
-	shared-pin;
-};
-
-&tpu {
-	pinctrl-0 = <&backlight_pins>;
-	pinctrl-names = "default";
-	status = "okay";
-};
-
-&usbphy {
-	status = "okay";
-	renesas,id-gpio = <&gpio4 6 GPIO_ACTIVE_LOW>;
-	renesas,vbus-gpio = <&gpio5 25 GPIO_ACTIVE_LOW>;
-	renesas,vbus-pwr-gpio = <&gpio5 24 GPIO_ACTIVE_LOW>;
-	renesas,wakeup;
-};
-
-&vin1 {
-	status = "okay";
-	pinctrl-0 = <&vin1_pins>;
-	pinctrl-names = "default";
-
-	port {
-		#address-cells = <1>;
-		#size-cells = <0>;
-
-		vin1ep: endpoint {
-			remote-endpoint = <&ov5640>;
-			bus-width = <8>;
-		};
-	};
 };
diff --git a/arch/arm/boot/dts/r8a7745-iwg22m-rev4.dtsi b/arch/arm/boot/dts/r8a7745-iwg22m-rev4.dtsi
new file mode 100644
index 0000000..6765099
--- /dev/null
+++ b/arch/arm/boot/dts/r8a7745-iwg22m-rev4.dtsi
@@ -0,0 +1,36 @@
+/*
+ * Device Tree Source for the iWave-RZG1E-G22M SODIMM SOM R4.0
+ *
+ * Copyright (C) 2021 Renesas Electronics Corp.
+ *
+ * This file is licensed under the terms of the GNU General Public License
+ * version 2.  This program is licensed "as is" without any warranty of any
+ * kind, whether express or implied.
+ */
+
+#include "r8a7745-iwg22m.dtsi"
+
+&qspi {
+	flash: flash@0 {
+		compatible = "issi,is25lp016d", "jedec,spi-nor";
+		spi-max-frequency = <133000000>;
+	};
+};
+
+&pfc {
+	/delete-node/ sd1;
+};
+
+&i2c3 {
+	/delete-node/ rtc@68;
+	rtc@51 {
+		compatible = "nxp,pcf85263";
+		reg = <0x51>;
+	};
+};
+
+&sdhi1 {
+	/delete-property/ pinctrl-0;
+	/delete-property/ pinctrl-names;
+	status = "disabled";
+};
-- 
2.7.4

